base/src/main/java/uk/ac/ebi/atlas/solr/analytics/fullanalytics/FullAnalyticsSearchService.java

```diff
@@ -0,0 +1,182 @@
+                .sorted(COMPARATOR_BY_SPECIFICITY_AND_NAIVE_DIFF_EXPRESSION)
+                .limit(resultSize)
+                .map(SpecificGeneProfile::geneId)
+                .collect(Collectors.toList());
+    }
+
+    private Set<SpecificGeneProfile> mapJsonBucketsWithAverageExpressionAndCountsToGeneProfiles(String json) {
+        Collection<Map<String, Object>> buckets = JsonPath.read(json, "$.facets.genes.buckets.*");
+
+        return buckets.stream()
+                .filter(bucket -> bucket.containsKey("avg_expression"))
+                .map(bucket ->
+                        SpecificGeneProfile.create(
+                                (String) bucket.get("val"),
+                                ((Number) bucket.get("count")).intValue(),
+                                ((Number) bucket.get("avg_expression")).doubleValue(),
+                                0, 0.0))
+                .collect(Collectors.toSet());
+    }
+
+    private Set<SpecificGeneProfile> mapJsonBucketsWithMaxExpressionAndCountsToGeneProfiles(String json) {
+        Collection<Map<String, Object>> buckets = JsonPath.read(json, "$.facets.genes.buckets.*");
+
+        return buckets.stream()
+                .filter(bucket -> bucket.containsKey("max_expression"))
+                .map(bucket ->
+                        SpecificGeneProfile.create(
+                                (String) bucket.get("val"),
+                                0, 0.0,
+                                ((Number) bucket.get("count")).intValue(),
+                                ((Number) bucket.get("max_expression")).doubleValue()))
+                .collect(Collectors.toSet());
+    }
+
+    private Set<SpecificGeneProfile> mergeGeneProfiles(
+            Collection<SpecificGeneProfile> geneProfilesInSelectedAssayGroupsWithSelectionData,
+            Collection<SpecificGeneProfile> geneProfilesInSelectedAssayGroupsWithSliceData) {
+
+        Map<String, SpecificGeneProfile> geneProfilesInSelectedAssayGroupsWithSliceDataByGeneId =
+                geneProfilesInSelectedAssayGroupsWithSliceData.stream()
+                        .collect(Collectors.toMap(SpecificGeneProfile::geneId, profile -> profile));
+
+        return geneProfilesInSelectedAssayGroupsWithSelectionData.stream()
+                .map(profile ->
+                        SpecificGeneProfile.merge(
+                                profile,
+                                geneProfilesInSelectedAssayGroupsWithSliceDataByGeneId.get(profile.geneId())))
+                .collect(Collectors.toSet());
+    }
+
+    @AutoValue
+    static abstract class SpecificGeneProfile {
+        static final Comparator<SpecificGeneProfile> COMPARATOR_BY_SPECIFICITY_AND_NAIVE_DIFF_EXPRESSION = (o1, o2) ->
+            ComparisonChain.start()
+                    .compare(
+                            o1.countsInSelection() + o1.countsOutsideSelection(),
```

Comment:
After AnalyticsDao is refactored to provide an interface we can work with, we can create beautiful, beautiful tests for this comparator - it's now quite hard for me to assert the correctness of this method, although it seems to be doing a couple of things I recall from elsewhere.